src/lib/vrf_evaluator/vrf_evaluator.ml

```diff
@@ -0,0 +1,408 @@
+                  ; vrf_result }
+              in
+              Some slot_won )
+    in
+    let rec find_winning_slot (consensus_time : Consensus_time.t) =
+      let slot = Consensus_time.slot consensus_time in
+      let global_slot = Consensus_time.to_global_slot consensus_time in
+      t.current_slot <- Some global_slot ;
+      let epoch' = Consensus_time.epoch consensus_time in
+      [%log info] "Slot in an epoch: $slot"
+        ~metadata:[("slot", Slot.to_yojson slot)] ;
+      if Epoch.(epoch' > epoch) then Deferred.unit
+      else
+        let start = Time.now () in
+        match%bind evaluate_vrf ~consensus_time with
+        | None ->
+            [%log info] "Did not win a slot, took $time ms"
+              ~metadata:
+                [("time", `Float Time.(Span.to_ms (diff (now ()) start)))] ;
+            find_winning_slot (Consensus_time.succ consensus_time)
+        | Some slot_won ->
+            [%log info] "Won a slot, took $time ms"
+              ~metadata:
+                [("time", `Float Time.(Span.to_ms (diff (now ()) start)))] ;
+            Queue.enqueue slots_won slot_won ;
+            find_winning_slot (Consensus_time.succ consensus_time)
+    in
+    find_winning_slot start_consensus_time
+
+  let create config =
+    let reset_reader, reset_writer =
+      Pipe_lib.Strict_pipe.create ~name:"Vrf-evaluator reset"
+        Pipe_lib.Strict_pipe.Synchronous
+    in
+    let t =
+      { config
+      ; last_checked_slot_and_epoch= Public_key.Compressed.Table.create ()
+      ; slots_won=
+          Queue.create
+            ~capacity:
+              (Global_slot.to_int config.consensus_constants.slots_per_epoch)
+            ()
+      ; current_slot= None
+      ; current_epoch= Epoch.zero
+      ; epoch_state= {block_producer_keys= []}
+      ; reset_writer }
+    in
+    don't_wait_for
+      (Strict_pipe.Reader.iter_without_pushback reset_reader
+         ~f:(fun epoch_data ->
+           [%log' info config.logger] "Resetting epoch data for epoch %d"
+             (Epoch.to_int epoch_data.epoch) ;
+           Queue.clear t.slots_won ;
+           t.current_slot <- None ;
+           t.current_epoch <- epoch_data.epoch ;
+           don't_wait_for (evaluate epoch_data t) )) ;
+    return t
+end
+
+module Functions = struct
+  type ('i, 'o) t =
+    'i Bin_prot.Type_class.t
+    * 'o Bin_prot.Type_class.t
+    * (Worker_state.t -> 'i -> 'o Deferred.t)
+
+  let create input output f : ('i, 'o) t = (input, output, f)
+
+  let set_new_epoch_state =
+    create Consensus.Data.Epoch_data_for_vrf.Stable.Latest.bin_t Unit.bin_t
+      (fun w e ->
+        let logger = w.config.logger in
+        let update () =
+          [%log info]
+            "Updating epoch data for the VRF evaluation for epoch $epoch"
+            ~metadata:[("epoch", Epoch.to_yojson e.epoch)] ;
+          Strict_pipe.Writer.write w.reset_writer e
+        in
+        if Epoch.(succ e.epoch > w.current_epoch) then (
```

Comment:
We do re-evaluate from the current slot (it is part of the epoch data passed by the daemon). https://github.com/MinaProtocol/mina/blob/f94285ee9ac948e93f41152501978be430b4f360/src/lib/consensus/proof_of_stake.ml#L2938 
Although, the block produced could be off of an old block if the daemon hasn't caught up yet which is same as the current behaviour. We'll need to implement https://github.com/MinaProtocol/mina/issues/4165 to avoid creating blocks during initial catchup